### PR TITLE
Pipeline HPO

### PR DESCRIPTION
**Pull request by jcapels**
I added hyperparameter optimization of models to the pipelines.

For that, I had to change the signature of the classes of hyperparameter optimization to follow the **fit** signature.

Now the class and method call is as follows:

```python
params_dict_svc = {"C": [1.0, 1.2, 0.8], "kernel": ['linear', 'poly', 'rbf']}
optimizer = HyperparameterOptimizerValidation(SVC,
                                              metric=Metric(accuracy_score),
                                              maximize_metric=True,
                                              n_iter_search=2,
                                              params_dict=params_dict_svc,
                                              model_type="sklearn")
best_svm, best_hyperparams, all_results = optimizer.fit(train_dataset=train_dataset, valid_dataset=valid_dataset)
```

I had to do this to implement HPO in pipelines generically.

Right now one can optimize the hyperparameters of a model within a pipeline as follows:

```python
from deepmol.parameter_optimization import HyperparameterOptimizerValidation

steps = [('basic standardizing', BasicStandardizer()),
         ('morgan fingerprints', MorganFingerprint(radius=2, size=1024)),
         ('low variance feature selection', LowVarianceFS(threshold=0.1)),
         ]

params_dict_svc = {"C": [1.0, 1.2, 0.8], "kernel": ['linear', 'poly', 'rbf']}
optimizer = HyperparameterOptimizerValidation(SVC, metric=Metric(accuracy_score),
                                          maximize_metric=True,
                                          cv=3,
                                          n_iter_search=2,
                                          params_dict=params_dict_svc,
                                          model_type="sklearn")
pipeline = Pipeline(steps=steps, path="DRD2", hpo=optimizer)
pipeline.fit(train, validation)
```

The last step of the pipeline will always be the best model. Then, one can access the best hyperparameters and all results as follows:

```python
from deepmol.parameter_optimization import HyperparameterOptimizerValidation

steps = [('basic standardizing', BasicStandardizer()),
         ('morgan fingerprints', MorganFingerprint(radius=2, size=1024)),
         ('low variance feature selection', LowVarianceFS(threshold=0.1)),
         ]

params_dict_svc = {"C": [1.0, 1.2, 0.8], "kernel": ['linear', 'poly', 'rbf']}
optimizer = HyperparameterOptimizerValidation(SVC, metric=Metric(accuracy_score),
                                          maximize_metric=True,
                                          cv=3,
                                          n_iter_search=2,
                                          params_dict=params_dict_svc,
                                          model_type="sklearn")
pipeline = Pipeline(steps=steps, path="DRD2", hpo=optimizer)
pipeline.fit(train, validation)

print(pipeline.hpo_all_results_)
print(pipeline.hpo_best_hyperparams_)
```

### CODE REVIEW

File: src/deepmol/parameter_optimization/deepchem_hyperparameter_optimization.py [hunk 1]
```diff
@@ -66,35 +63,37 @@ def __init__(self,
...
-        if self.mode != dataset.mode:
+        if self.mode != train_dataset.mode:
             raise ValueError(f'The mode of the model and the dataset must be the same. Got {self.mode} and '
-                             f'{dataset.mode} respectively.')
+                             f'{train_dataset.mode} respectively.')
         results_dict = defaultdict(list)
 
         # split dataset into folds
-        if dataset.mode == 'classification':
+        if train_dataset.mode == 'classification':
```
Comment (jcorreia11) [inline src/deepmol/parameter_optimization/deepchem_hyperparameter_optimization.py:83]: maybe later we can add support for other splitters and for multitask.

File: src/deepmol/pipeline/pipeline.py [hunk 2]
```diff
@@ -121,13 +131,33 @@ def _fit(self, dataset: Dataset) -> 'Pipeline':
...
+                        validation_dataset = transformer.transform(validation_dataset)
+            else:
+                for name, transformer in self.steps:
+                    train_dataset = transformer.fit_transform(train_dataset)
+                    if validation_dataset is not None:
+                        validation_dataset = transformer.transform(validation_dataset)
+
+            best_model, self.hpo_best_hyperparams_, self.hpo_all_results_ = self.hpo.fit(train_dataset,
+                                                                                         validation_dataset)
+            if self.is_prediction_pipeline():
```
Comment (jcorreia11) [inline src/deepmol/pipeline/pipeline.py:156]: we need to assure that refit=True

**Review by jcorreia11 — APPROVED** [review]

**Comment by jcorreia11** [comment]
Nice addition to the Pipeline. I am approving it. Just have one suggestion of improvement for the future and one small concern regarding the addition of the best model to the pipeline if refit is set to False. Check the comments and let me know what you think. After that, you can merge the PR. Good job!

File: src/deepmol/pipeline/pipeline.py (hunk 2 above)
Comment (jcapels) [reply src/deepmol/pipeline/pipeline.py:156]: I checked the code and it seems that if the user sets refit=False, it maintains the best_estimator of the CV, whereas if it is True, it retrains the model with the whole dataset... I think it is the expected behaviour, no?

File: src/deepmol/parameter_optimization/deepchem_hyperparameter_optimization.py (hunk 1 above)
Comment (jcapels) [reply src/deepmol/parameter_optimization/deepchem_hyperparameter_optimization.py:83]: I will put it in an issue

File: src/deepmol/pipeline/pipeline.py (hunk 2 above)
Comment (jcorreia11) [reply src/deepmol/pipeline/pipeline.py:156]: Great. That's it.

**Pull request merged**